File: Euclid/Game/Player/Player.cs
using Euclid.Game.Entity.Humanoid;
using Euclid.Messages;
using Euclid.Messages.Outgoing;
using Euclid.Network.Session;
using Euclid.Storage.Database.Access;
using Euclid.Storage.Database.Data;
using log4net;
using System;
using System.Linq;
using System.Reflection;

namespace Euclid.Game
{
    public class Player : Humanoid
    {
        #region Fields

        private ILog log = LogManager.GetLogger(typeof(Player));
        private PlayerData playerData;

        #endregion

        #region Interface properties

        /// <summary>
        /// Get entity data
        /// </summary>
        public override IEntityData EntityData => (IEntityData)playerData;

        /// <summary>
        /// Get entity type
        /// </summary>
        public override EntityType EntityType => EntityType.PLAYER;

        /// <summary>
        /// Get room entity
        /// </summary>
        public override RoomEntity RoomEntity { get; set; }

        #endregion

        #region Properties

        /// <summary>
        /// Get the connection session
        /// </summary>
        public ConnectionSession Connection { get; private set; }


        public ConnectionMode ConnectionMode => Connection.Mode;

        /// <summary>
        /// Get the logging
        /// </summary>
        public ILog Log => log;

        /// <summary>
        /// Get entity data
        /// </summary>
        public PlayerData Details => playerData;

        /// <summary>
        /// Get inventory instance for user
        /// </summary>
        public Inventory Inventory { get; private set; }

        /// <summary>
        /// Whether the player has logged in or not
        /// </summary>
        public bool Authenticated { get; private set; }

        /// <summary>
        /// The time when player connected
        /// </summary>
        public DateTime AuthenticationTime { get; private set; }

        /// <summary>
        /// Get user group
        /// </summary>
        public UserGroup UserGroup => PermissionsManager.Instance.Ranks[Details.Rank];

        /// <summary>
        /// Get whether disconnect was handled
        /// </summary>
        public bool Disconnected { get; private set; }

        /// <summary>
        /// Get the player by the main server connection
        /// </summary>
        public Player MainServer => PlayerManager.Instance.Players.FirstOrDefault(x => x.Details.Id == Details.Id && x.ConnectionMode == ConnectionMode.MAIN);

        /// <summary>
        /// Get the player by the private server connection
        /// </summary>
        public Player PrivateServer => PlayerManager.Instance.Players.FirstOrDefault(x => x.Details.Id == Details.Id && x.ConnectionMode == ConnectionMode.PRIVATE);

        /// <summary>
        /// Get the player by the public server connection
        /// </summary>
        public Player PublicServer => PlayerManager.Instance.Players.FirstOrDefault(x => x.Details.Id == Details.Id && x.ConnectionMode == ConnectionMode.PUBLIC);

        #endregion

        #region Constructors

        /// <summary>
        /// Constructor for player.
        /// </summary>
        /// <param name="channel">the channel</param>
        public Player(ConnectionSession connectionSession)
        {
            Connection = connectionSession;
            playerData = new PlayerData();
            RoomEntity = new RoomPlayer(this);

            if (ConnectionMode == ConnectionMode.MAIN)
                log = LogManager.GetLogger(Assembly.GetExecutingAssembly(), $"[Main] Connection {connectionSession.Channel.Id}");

            if (ConnectionMode == ConnectionMode.PRIVATE)
                log = LogManager.GetLogger(Assembly.GetExecutingAssembly(), $"[Private] Connection {connectionSession.Channel.Id}");

            if (ConnectionMode == ConnectionMode.PUBLIC)
                log = LogManager.GetLogger(Assembly.GetExecutingAssembly(), $"[Public] Connection {connectionSession.Channel.Id}");
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Login handler
        /// </summary>
        /// <param name="ssoTicket">the sso ticket</param>
        /// <returns></returns>
        public bool TryLogin(string username, string password)
        {
            playerData = UserDao.Login(username, password);

            if (playerData == null)
            {
                Send(new SYSTEMBROADCAST("Your username or password was incorrect."));
                return false;
            }

            if (ConnectionMode == ConnectionMode.MAIN)
                log = LogManager.GetLogger(Assembly.GetExecutingAssembly(), $"[Main] Player {playerData.Name}");

            if (ConnectionMode == ConnectionMode.PRIVATE)
                log = LogManager.GetLogger(Assembly.GetExecutingAssembly(), $"[Private] Player {playerData.Name}");

            if (ConnectionMode == ConnectionMode.PUBLIC)
                log = LogManager.GetLogger(Assembly.GetExecutingAssembly(), $"[Public] Player {playerData.Name}");

            log.Debug($"Player {playerData.Name} has logged in");

            playerData.LastOnline = DateTime.Now;
            
            Inventory = new Inventory(this);

            // Don't bother loading inventory on anything but the private connection
            if (this.ConnectionMode == ConnectionMode.PRIVATE)
            {
                Inventory.Load();
            }
            UserDao.SaveLastOnline(playerData);
            PlayerManager.Instance.AddPlayer(this);

            Authenticated = true;
            AuthenticationTime = DateTime.Now;

            return true;
        }

        /// <summary>
        /// Send message composer
        /// </summary>
        public void Send(IMessageComposer composer)
        {
            Connection.Send(composer);
        }

        /// <summary>
        /// Disconnection handler
        /// </summary>
        public void OnDisconnect()
        {
            if (!Authenticated)
                return;

            PlayerManager.Instance.RemovePlayer(this);

            playerData.LastOnline = DateTime.Now;
            UserDao.SaveLastOnline(playerData);

            if (RoomEntity.Room != null)
            {
                if (RoomEntity.BeingKicked)
                {
                    // Kick instead of leave room
                    RoomEntity.kick(false);
                }
                else
                {
                    RoomEntity.Room.EntityManager.LeaveRoom(this);
                }
            }

            this.Disconnected = true;

            /*long timeInSeconds = (long)(DateTime.Now - AuthenticationTime).TotalSeconds;
            settings.OnlineTime += timeInSeconds;*/
           
        }

        public bool Disconnect() {

            Connection.Disconnect();
            return true;
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/game/Player.java
package euclid.game;

import euclid.game.entity.Humanoid;
import euclid.game.entity.EntityType;
import euclid.game.entity.RoomEntity;
import euclid.game.entity.RoomPlayer;
import euclid.game.item.Inventory;
import euclid.game.permissions.PermissionsManager;
import euclid.game.permissions.UserGroup;
import euclid.messages.IMessageComposer;
import euclid.messages.outgoing.SYSTEMBROADCAST;
import euclid.network.session.ConnectionMode;
import euclid.network.session.ConnectionSession;
import euclid.storage.database.access.UserDao;
import euclid.storage.database.data.PlayerData;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.time.LocalDateTime;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class Player extends Humanoid {

    private Logger log;
    private final ConnectionSession connection;
    private PlayerData playerData;
    private RoomEntity roomEntity;
    private Inventory inventory;
    private boolean authenticated;
    private LocalDateTime authenticationTime;
    private boolean disconnected;

    public Player(ConnectionSession connectionSession) {
        this.connection = connectionSession;
        this.playerData = new PlayerData();
        this.roomEntity = new RoomPlayer(this);
        this.entityData = playerData;
        this.entityType = EntityType.PLAYER;

        if (getConnectionMode() == ConnectionMode.MAIN)
            log = LogManager.getLogger("[Main] Connection " + connectionSession.getChannel().id().asLongText());
        else if (getConnectionMode() == ConnectionMode.PRIVATE)
            log = LogManager.getLogger("[Private] Connection " + connectionSession.getChannel().id().asLongText());
        else
            log = LogManager.getLogger("[Public] Connection " + connectionSession.getChannel().id().asLongText());
    }

    @Override
    public PlayerData getEntityData() {
        return playerData;
    }

    @Override
    public EntityType getEntityType() {
        return EntityType.PLAYER;
    }

    @Override
    public RoomEntity getRoomEntity() {
        return roomEntity;
    }

    @Override
    public void setRoomEntity(RoomEntity roomEntity) {
        this.roomEntity = roomEntity;
    }

    public ConnectionSession getConnection() {
        return connection;
    }

    public ConnectionMode getConnectionMode() {
        return connection.getMode();
    }

    public Logger getLog() {
        return log;
    }

    public PlayerData getDetails() {
        return playerData;
    }

    public void setDetails(PlayerData details) {
        this.playerData = details;
        this.entityData = details;
    }

    public Inventory getInventory() {
        return inventory;
    }

    public void setInventory(Inventory inventory) {
        this.inventory = inventory;
    }

    public boolean isAuthenticated() {
        return authenticated;
    }

    public void setAuthenticated(boolean authenticated) {
        this.authenticated = authenticated;
    }

    public LocalDateTime getAuthenticationTime() {
        return authenticationTime;
    }

    public void setAuthenticationTime(LocalDateTime authenticationTime) {
        this.authenticationTime = authenticationTime;
    }

    public UserGroup getUserGroup() {
        return PermissionsManager.getInstance().getRanks().get(playerData.getRank());
    }

    public boolean isDisconnected() {
        return disconnected;
    }

    public void setDisconnected(boolean disconnected) {
        this.disconnected = disconnected;
    }

    public Player getMainServer() {
        List<Player> players = PlayerManager.getInstance().getPlayers();
        return players.stream()
                .filter(x -> x.getDetails().getId() == playerData.getId() && x.getConnectionMode() == ConnectionMode.MAIN)
                .findFirst()
                .orElse(null);
    }

    public Player getPrivateServer() {
        List<Player> players = PlayerManager.getInstance().getPlayers();
        return players.stream()
                .filter(x -> x.getDetails().getId() == playerData.getId() && x.getConnectionMode() == ConnectionMode.PRIVATE)
                .findFirst()
                .orElse(null);
    }

    public Player getPublicServer() {
        List<Player> players = PlayerManager.getInstance().getPlayers();
        return players.stream()
                .filter(x -> x.getDetails().getId() == playerData.getId() && x.getConnectionMode() == ConnectionMode.PUBLIC)
                .findFirst()
                .orElse(null);
    }

    public boolean tryLogin(String username, String password) {
        playerData = UserDao.login(username, password);

        if (playerData == null) {
            send(new SYSTEMBROADCAST("Your username or password was incorrect."));
            return false;
        }

        if (getConnectionMode() == ConnectionMode.MAIN)
            log = LogManager.getLogger("[Main] Player " + playerData.getName());
        else if (getConnectionMode() == ConnectionMode.PRIVATE)
            log = LogManager.getLogger("[Private] Player " + playerData.getName());
        else
            log = LogManager.getLogger("[Public] Player " + playerData.getName());

        log.debug("Player {} has logged in", playerData.getName());

        playerData.setLastOnline(LocalDateTime.now());
        this.entityData = playerData;

        inventory = new Inventory(this);

        if (getConnectionMode() == ConnectionMode.PRIVATE) {
            inventory.load();
        }

        UserDao.saveLastOnline(playerData);
        PlayerManager.getInstance().addPlayer(this);

        authenticated = true;
        authenticationTime = LocalDateTime.now();

        return true;
    }

    public void send(IMessageComposer composer) {
        connection.send(composer);
    }

    public void onDisconnect() {
        if (!authenticated)
            return;

        PlayerManager.getInstance().removePlayer(this);

        playerData.setLastOnline(LocalDateTime.now());
        UserDao.saveLastOnline(playerData);

        if (roomEntity.getRoom() != null) {
            if (roomEntity.isBeingKicked()) {
                roomEntity.kick(false);
            } else {
                roomEntity.getRoom().getEntityManager().leaveRoom(this);
            }
        }

        this.disconnected = true;
    }

    public boolean disconnect() {
        connection.disconnect();
        return true;
    }
}
